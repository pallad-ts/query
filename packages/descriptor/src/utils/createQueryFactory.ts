import { z } from "zod";
import { fromNullable } from "@sweet-monads/maybe";
import { QueryDescriptor } from "../QueryDescriptor";
import { fromZodValidation } from "@pallad/violations-zod";
import { right } from "@sweet-monads/either";

export function createQueryFactory(
	config: QueryDescriptor.Config,
	filtersValidator: QueryDescriptor.FiltersValidator<any> | undefined
) {
	const paginationSchema = getPaginationSchema(config.pagination);
	const sortingSchema = getSortingSchema(config.sorting);

	let schema = z.object({
		filters: z
			.object({})
			.passthrough()
			.default(() => ({})),
	});

	schema = paginationSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);
	schema = sortingSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);

	return (input: unknown) => {
		return fromZodValidation(schema, input).chain(({ filters, ...meta }) => {
			if (filtersValidator) {
				return filtersValidator(filters).map(filters => ({ ...meta, filters }));
			}
			return right({ ...meta, filters });
		});
	};
}

function createSortableFieldSchema(allowedFields: readonly [string, ...string[]]) {
	return z.object({
		field: z.enum(allowedFields),
		direction: z.enum(["DESC", "ASC"]),
	});
}

function getPaginationSchema(pagination: QueryDescriptor.Pagination | undefined) {
	return fromNullable(pagination).map(x => {
		if (x.type === "byCursor") {
			return z.object({
				limit: getLimitSchema(x),
				after: z.string().optional(),
				before: z.string().optional(),
			});
		}
		return z.object({
			limit: getLimitSchema(x),
			offset: z.number().int().min(0).default(0),
		});
	});
}

function getLimitSchema(pagination: QueryDescriptor.Pagination) {
	return z.number().int().min(1).max(pagination.maxLimit).default(pagination.defaultLimit);
}

function getSortingSchema(sorting: QueryDescriptor.Sorting<any> | undefined) {
	return fromNullable(sorting).map(x => {
		const sortByType = createSortableFieldSchema(x.sortableFields);
		if (x.type === "single") {
			return z.object({
				sortBy: sortByType.default(x.defaultSorting),
			});
		}

		return z.object({
			sortBy: z
				.array(sortByType)
				.nonempty()
				.default(x.defaultSorting as [any, ...any[]]),
		});
	});
}
