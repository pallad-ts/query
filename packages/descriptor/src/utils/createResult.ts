import { fromNullable, Maybe } from "@sweet-monads/maybe";
import {
	PaginableByCursor,
	PaginableByOffset,
	Query,
	SortableMulti,
	SortableSingle,
} from "@pallad/query";
import { QueryDescriptor } from "../QueryDescriptor";

export function createResult(
	config: QueryDescriptor.Config,
	resultOptions: QueryDescriptor.ResultOptions<any, any, any>
) {
	let meta: any = {};
	composeMetaForPagination(config.pagination, resultOptions).chain(x =>
		Object.assign(meta, x)
	);
	composeMetaForSorting(config.sorting, resultOptions.query).chain(x => Object.assign(meta, x));

	const hasMeta = Object.keys(meta).length > 0;

	return {
		results: resultOptions.results,
		...(hasMeta ? { meta } : {}),
	};
}

function composeMetaForPagination(
	pagination: QueryDescriptor.Pagination | undefined,
	options: QueryDescriptor.ResultOptions<any, any, any>
) {
	return fromNullable(pagination).map(settings => {
		if (settings.type === "byOffset") {
			const finalQuery = options.query as Required<PaginableByOffset>;
			return {
				limit: finalQuery.limit,
				offset: finalQuery.offset,
			} as PaginableByOffset.ResultMeta;
		}

		const finalQuery = options.query as Required<PaginableByCursor>;
		return {
			limit: finalQuery.limit,
			nextPage: "nextPageCursor" in options ? options.nextPageCursor : undefined,
			previousPage: "previousPageCursor" in options ? options.previousPageCursor : undefined,
		} as PaginableByCursor.ResultMeta;
	});
}

function composeMetaForSorting(
	sorting: QueryDescriptor.Sorting<any> | undefined,
	query: Query<any> & (SortableMulti<any> | SortableSingle<any>)
) {
	return fromNullable(sorting).map(() => {
		return { sortBy: query.sortBy } as
			| SortableSingle.ResultMeta<any>
			| SortableMulti.ResultMeta<any>;
	});
}
