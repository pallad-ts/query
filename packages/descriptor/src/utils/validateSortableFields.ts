import { ERRORS } from "../errors";

export function validateSortableFields<T>(fields: T[]): asserts fields is [T, ...T[]] {
	if (fields.length === 0) {
		throw ERRORS.SORTABLE_FIELDS_LIST_CANNOT_BE_EMPTY.create();
	}
}
