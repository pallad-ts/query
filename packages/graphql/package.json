{
  "name": "@pallad/query-graphql",
  "version": "2.0.1",
  "description": "Helper for graphql to create @pallad/query related types",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "scripts": {
    "test": "pallad-scripts test",
    "lint": "pallad-scripts lint",
    "compile": "pallad-scripts compile",
    "prepublishOnly": "npm run compile"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pallad-ts/query.git",
    "directory": "packages/main"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl/)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pallad-ts/query/issues"
  },
  "homepage": "https://github.com/pallad-ts/query#readme",
  "peerDependencies": {
    "graphql": "^16.3.0",
    "graphql-compose": "^9.0.8"
  },
  "dependencies": {
    "@pallad/builder": "^1.0.1",
    "@pallad/query": "^2.0.0",
    "predicates": "^2.0.3"
  }
}
