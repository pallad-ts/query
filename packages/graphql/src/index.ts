export * from './types';
export * from './createInputSortType';
export * from './createResultListType';
export * from './createSortFieldType';
export * from './createSortTypeBase';
export * from './GraphQLQueryBuilder';
