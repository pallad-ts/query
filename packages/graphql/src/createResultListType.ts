import {GraphQLList, GraphQLNonNull, GraphQLObjectType} from "graphql";
import {ObjectTypeComposer} from "graphql-compose";

export function createResultListType(entityType: GraphQLObjectType | ObjectTypeComposer) {
	if (entityType instanceof GraphQLObjectType) {
		return new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(entityType)));
	}
	return entityType.NonNull.getTypePlural().NonNull;
}
