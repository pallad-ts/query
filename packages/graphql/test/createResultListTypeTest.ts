import {GraphQLObjectType} from "graphql/type";
import {GraphQLString} from "graphql";
import {createResultListType} from "@src/createResultListType";
import {ObjectTypeComposer} from "graphql-compose";
import {printObject} from "graphql-compose/lib/utils/schemaPrinter";

describe('createResultListType', () => {
	function assertValidType(type: any) {
		const objectType = ObjectTypeComposer.createTemp({
			name: 'BaseWrapper',
			fields: {
				results: type
			}
		});

		expect(printObject(objectType.getType()))
			.toMatchSnapshot();
	}

	it('from raw GraphQL type', () => {
		const entity = new GraphQLObjectType({
			name: 'Base',
			fields: {
				foo: {type: GraphQLString}
			}
		});

		const result = createResultListType(entity);
		assertValidType(result);
	});

	it('from type composer', () => {
		const entity = ObjectTypeComposer.createTemp({
			name: 'Base',
			fields: {
				foo: GraphQLString
			}
		});

		const result = createResultListType(entity);
		assertValidType(result);
	});
});
