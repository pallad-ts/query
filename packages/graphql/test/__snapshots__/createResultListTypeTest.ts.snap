// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResultListType from raw GraphQL type 1`] = `
"type BaseWrapper {
  results: [Base!]!
}"
`;

exports[`createResultListType from type composer 1`] = `
"type BaseWrapper {
  results: [Base!]!
}"
`;
